rpcs3/Emu/Cell/SPUThread.cpp

```diff
@@ -661,34 +680,142 @@ void SPUThread::do_dma_transfer(const spu_mfc_cmd& args, bool from_mfc)
+	for (u32 i = 0; i < mfc_size; i++)
+	{
+		if ((mfc_queue[i].cmd & ~0xc) == MFC_BARRIER_CMD)
+		{
+			return false;
+		}
 
-	const auto mfc = fxm::check_unlocked<mfc_thread>();
+		if (mfc_queue[i].tag == ch_mfc_cmd.tag)
+		{
+			if (ch_mfc_cmd.cmd & MFC_FENCE_MASK || mfc_queue[i].cmd & MFC_BARRIER_MASK)
+			{
+				return false;
+			}
+		}
+	}
 
-	// Check queue size
-	auto check_queue_size = [&]()
+	return true;
+}
+
+bool SPUThread::do_list_transfer(spu_mfc_cmd& args)
+{
+	vm::reader_lock lock;
+
+	struct list_element
+	{
+		be_t<u16> sb; // Stall-and-Notify bit (0x8000)
+		be_t<u16> ts; // List Transfer Size
+		be_t<u32> ea; // External Address Low
+	} item{};
+
+	while (args.size)
 	{
-		while (mfc_queue.size() >= 16)
+		if (UNLIKELY(item.sb & 0x8000))
 		{
-			if (test(state, cpu_flag::stop + cpu_flag::dbg_global_stop))
+			ch_stall_mask |= (1u << args.tag);
+			ch_stall_stat.push_or(*this, 1u << args.tag);
+			ch_event_stat |= SPU_EVENT_SN;
+			return false;
+		}
+
+		args.lsa &= 0x3fff0;
+		item = _ref<list_element>(args.eal & 0x3fff8);
+
+		const u32 size = item.ts;
+		const u32 addr = item.ea;
+
+		if (size)
+		{
+			if (!vm::check_addr(addr, size, vm::page_readable | (args.cmd & MFC_PUT_CMD ? vm::page_writable : 0)))
 			{
-				return;
+				LOG_FATAL(SPU, "Access violation %s location 0x%x (%s, size=0x%x)",
+					args.cmd & MFC_PUT_CMD ? "writing" : "reading", addr, args.cmd, size);
+				return false;
 			}
 
-			// TODO: investigate lost notifications
-			std::this_thread::yield();
-			_mm_lfence();
+			spu_mfc_cmd transfer;
+			transfer.eal  = addr;
+			transfer.eah  = 0;
+			transfer.lsa  = args.lsa | (addr & 0xf);
+			transfer.tag  = args.tag;
+			transfer.cmd  = MFC(args.cmd & ~MFC_LIST_MASK);
+			transfer.size = size;
+
+			do_dma_transfer(transfer);
+			const u32 add_size = std::max<u32>(size, 16);
+			args.lsa += add_size;
 		}
-	};
+
+		args.eal += 8;
+		args.size -= 8;
+	}
+
+	return true;
+}
+
+bool SPUThread::do_lluc(const spu_mfc_cmd& args)
+{
+	const u32 addr = args.eal;
+	auto& data = vm::_ref<decltype(rdata)>(addr);
+	const auto to_write = _ref<decltype(rdata)>(args.lsa & 0x3ffff);
+
+	vm::reservation_acquire(addr, 128);
+
+	// Store unconditionally
+	if (s_use_rtm && utils::transaction_enter())
+	{
+		if (!vm::reader_lock{ vm::try_to_lock })
+		{
+			_xabort(0);
+		}
+
+		data = to_write;
+		vm::reservation_update(addr, 128);
+		vm::notify(addr, 128);
+		_xend();
+	}
+	else
+	{
+		vm::writer_lock lock(0);
+		data = to_write;
+		vm::reservation_update(addr, 128);
+		vm::notify(addr, 128);
+	}
+
+	return true;
+}
+
+u32 SPUThread::get_mfc_completed()
+{
+	u32 completed = ch_tag_mask;
+
+	for (u32 i = 0; i < mfc_size; i++)
+	{
+		if ((mfc_queue[i].cmd & ~0xc) != MFC_BARRIER_CMD)
+		{
+			completed &= ~(1u << mfc_queue[i].tag);
+		}
+	}
+
+	return completed;
+}
+
+void SPUThread::process_mfc_cmd()
+{
+	verify("MFC Queue Size" HERE), mfc_size < 16;
```

Comment:
when the queue is full (size is 16) the command needs to wait until theres a free place for it, so put here a `while (mfc_size == 16)` instead and process mfc commands inside the loop